File: src/services/session.ts
import { invoke, type Channel } from '@tauri-apps/api/core'

import type {
  AgentEvent,
  ApprovalDecision,
  SendMessageInput,
  SessionDetail,
  SessionInfo,
  StartSessionInput,
  TurnHandle,
} from '../types/chat'

/** 会话列表；缺省返回全部 Agent 的会话，按更新时间倒序。 */
export function listSessions(agentId?: number): Promise<SessionInfo[]> {
  return invoke<SessionInfo[]>('session_list', { agentId: agentId ?? null })
}

/** 会话 + 全部消息。 */
export function getSession(sessionId: number): Promise<SessionDetail> {
  return invoke<SessionDetail>('session_get', { sessionId })
}

export function renameSession(sessionId: number, title: string): Promise<SessionInfo> {
  return invoke<SessionInfo>('session_rename', { sessionId, title })
}

export function deleteSession(sessionId: number): Promise<void> {
  return invoke<void>('session_delete', { sessionId })
}

/** 新建会话并开始第一轮；进度经 `onEvent` 推送。 */
export function startSession(
  input: StartSessionInput,
  onEvent: Channel<AgentEvent>,
): Promise<TurnHandle> {
  return invoke<TurnHandle>('agent_start_session', { input, onEvent })
}

/** 在既有会话上开始新一轮；进度经 `onEvent` 推送。 */
export function sendMessage(
  input: SendMessageInput,
  onEvent: Channel<AgentEvent>,
): Promise<TurnHandle> {
  return invoke<TurnHandle>('agent_send_message', { input, onEvent })
}

export function cancelTurn(sessionId: number): Promise<void> {
  return invoke<void>('agent_cancel_turn', { sessionId })
}

export function approveToolCall(
  sessionId: number,
  turnId: string,
  callId: string,
  decision: ApprovalDecision,
): Promise<void> {
  return invoke<void>('agent_approve_tool_call', { sessionId, turnId, callId, decision })
}

File: src-tauri/src/controller/session.rs
//! 会话与轮次命令；仅做参数透传，把 `Channel<AgentEvent>` 包装为 `ChannelEmitter`。

use std::sync::Arc;

use tauri::ipc::Channel;

use crate::interface::error::ApiError;
use crate::interface::event::AgentEvent;
use crate::interface::session::{
    ApprovalDecision, SendMessageInput, SessionDetail, SessionInfo, StartSessionInput, TurnHandle,
};
use crate::service::event::{ChannelEmitter, EventEmitter};
use crate::AppState;

#[tauri::command]
pub fn session_list(
    state: tauri::State<'_, AppState>,
    agent_id: Option<i64>,
) -> Result<Vec<SessionInfo>, ApiError> {
    state.session.list_sessions(agent_id)
}

#[tauri::command]
pub fn session_get(
    state: tauri::State<'_, AppState>,
    session_id: i64,
) -> Result<SessionDetail, ApiError> {
    state.session.get_session(session_id)
}

#[tauri::command]
pub fn session_rename(
    state: tauri::State<'_, AppState>,
    session_id: i64,
    title: String,
) -> Result<SessionInfo, ApiError> {
    state.session.rename_session(session_id, &title)
}

#[tauri::command]
pub fn session_delete(state: tauri::State<'_, AppState>, session_id: i64) -> Result<(), ApiError> {
    state.session.delete_session(session_id)
}

#[tauri::command]
pub fn agent_start_session(
    state: tauri::State<'_, AppState>,
    input: StartSessionInput,
    on_event: Channel<AgentEvent>,
) -> Result<TurnHandle, ApiError> {
    let emitter: Arc<dyn EventEmitter> = Arc::new(ChannelEmitter::new(on_event));
    state.session.start_session(input, emitter)
}

#[tauri::command]
pub fn agent_send_message(
    state: tauri::State<'_, AppState>,
    input: SendMessageInput,
    on_event: Channel<AgentEvent>,
) -> Result<TurnHandle, ApiError> {
    let emitter: Arc<dyn EventEmitter> = Arc::new(ChannelEmitter::new(on_event));
    state.session.send_message(input, emitter)
}

#[tauri::command]
pub fn agent_cancel_turn(
    state: tauri::State<'_, AppState>,
    session_id: i64,
) -> Result<(), ApiError> {
    state.session.cancel_turn(session_id)
}

#[tauri::command]
pub fn agent_approve_tool_call(
    state: tauri::State<'_, AppState>,
    session_id: i64,
    turn_id: String,
    call_id: String,
    decision: ApprovalDecision,
) -> Result<(), ApiError> {
    state
        .session
        .approve_tool_call(session_id, &turn_id, &call_id, decision)
}

#[tauri::command]
pub fn agent_answer_question(
    state: tauri::State<'_, AppState>,
    session_id: i64,
    turn_id: String,
    call_id: String,
    answer: String,
) -> Result<(), ApiError> {
    state
        .session
        .answer_question(session_id, &turn_id, &call_id, &answer)
}

#[tauri::command]
pub fn agent_skip_question(
    state: tauri::State<'_, AppState>,
    session_id: i64,
    turn_id: String,
    call_id: String,
) -> Result<(), ApiError> {
    state.session.skip_question(session_id, &turn_id, &call_id)
}
